push/sched/scheduler.go

```
@@ -45,93 +62,107 @@ func (s *ShardScheduler) Stop() {
-		if holder.errChan != nil {
-			holder.errChan <- err
-		} else if err != nil {
-			log.Errorf("Failed to execute action: %+v", err)
+func (s *ShardScheduler) WaitForProcessingToComplete(ch chan struct{}) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	lastQueued := s.lastQueuedReceiverSeq
+	go func() {
+		start := time.Now()
+		for {
+			lastProcessed := atomic.LoadUint64(&s.lastProcessedReceiverSeq)
+			if lastProcessed == lastQueued {
+				ch <- struct{}{}
+				return
+			}
+			time.Sleep(1 * time.Millisecond)
+			if time.Now().Sub(start) > 10*time.Second {
+				log.Warnf("timed out waiting for shard %d processing to complete", s.shardID)
+				ch <- struct{}{}
+				return
+			}
 		}
-	}
-}
-
-func (s *ShardScheduler) ScheduleAction(action Action) chan error {
-	// Channel size is 1 - we don't want writer to block waiting for reader
-	ch := make(chan error, 1)
-	s.submitAction(&actionHolder{
-		action:  action,
-		errChan: ch,
-	})
-	return ch
-}
-
-func (s *ShardScheduler) ScheduleActionFireAndForget(action Action) {
-	s.submitAction(&actionHolder{
-		action: action,
-	})
+	}()
 }
 
 func (s *ShardScheduler) ShardID() uint64 {
 	return s.shardID
 }
 
-func (s *ShardScheduler) submitAction(action *actionHolder) {
+func (s *ShardScheduler) getRowBatch() ([]cluster.ForwardRow, bool) {
 	s.lock.Lock()
-	if !s.started {
-		return
+	defer s.lock.Unlock()
+
+	numRows := len(s.forwardRows)
+	if numRows > maxProcessBatchRows {
+		numRows = maxProcessBatchRows
+	}
+	rows := s.forwardRows[:numRows]
+	s.forwardRows = s.forwardRows[numRows:]
+	more := len(s.forwardRows) > 0
+	if !more {
+		s.processorRunning = false
+	}
+	return rows, more
+}
+
+func (s *ShardScheduler) runLoop() {
+	for {
+		_, ok := <-s.actions
+		if !ok {
+			break
+		}
+		rows, more := s.getRowBatch()
+		if len(rows) == 0 {
+			panic("expected rows")
+		}
+		if err := s.batchHandler.HandleBatch(s.shardID, rows); err != nil {
+			log.Errorf("failed to process batch: %+v", err)
+			return
```

Should we return the error here?

---

I don't think there's any obvious place to propagate it to